### PR TITLE
fixes #4315 Return standard error responses for host power management vi...

### PR DESCRIPTION
**Pull request by mchesler**
...a API

### CODE REVIEW

**Comment by skottler** [comment]
[test]

**Comment by domcleal** [comment]
@eLobato could you review please?

**Comment by dLobatog** [comment]
@mchesler 
:+1: +1 to standardized error messages, good catch!

However, I don't think returning `@host.power.send...` is the most sensible option here. Lets go over the two possible cases.

`true` - API returns the whole host object, because that's the 'standard way'. There's no real information to know that the call went well other than the 200.. it makes the API response look identical to a /host/id PUT response and what happened is unclear to the developer.

`false` -  This returns

``` json
{  "error": {"message":"resource have no errors"} }
```

This is arguably not informative at all. At best you could make another if/else clause with the value returned from `@host.power.send`, but I think a simple

``` json
{  "power" : "reboot" }
```

or 

``` json
{  "power" : "error rebooting" } 
```

In this case the messages ("reboot" and "error rebooting") come straight from the host where you issued the command, and the developer can easily tell what happened.

@isratrade might as well comment on this as I could be biased.

**Comment by isratrade** [comment]
@eLobato, before and after this PR, `@host.power.send` is returned.  This PR just appears to deal with formatting?

**Comment by dLobatog** [comment]
@isratrade `process_response @host.power.send(params[:power_action])` returns the whole `Host` object if successful.

**Comment by mchesler** [comment]
@eLobato It seems like the most sensible approach would be to return the hash that comes back from the `@host.power.send(params[:power_action])` call.  As an example, a power on call for vmware returns a hash with a single key, `task_state` , which could have values `success`, `running`, `queued` or `error` (see https://github.com/fog/fog/blob/master/lib/fog/vsphere/requests/compute/vm_power_on.rb#L15).  Please let me know if my understanding is completely out of whack here.

**Comment by dLobatog** [comment]
Alright, admittedly I've only tried with Openstack and libvirt, which just returned the whole Host object. In the case of Openstack, it just returns the result of another request https://github.com/fog/fog/blob/master/lib/fog/openstack/requests/compute/server_action.rb
which in turns returns..
http://docs.openstack.org/api/openstack-compute/2/content/POST_resume_v2__tenant_id__servers__server_id__action_ext-os-admin-actions.html

>  This operation does not return a response body

for resume, pause, suspend, reboot, etc.

Libvirt does the same and returns a very useful..
http://libvirt.org/html/libvirt-libvirt.html#virDomainResume

> 0 in case of success and -1 in case of failure.

However I just saw that EC2 and VSphere return much more sensible things, such as the hash you posted or EC2's hash which says what's the current, and previous state. https://github.com/fog/fog/blob/master/lib/fog/aws/requests/compute/start_instances.rb#L46

In short, **thank you** for bringing up this issue as we'll have to look into how to properly respond to all Compute Resources we support. 

My preference is to keep responding `"power" : "on"` to avoid breaking already existing programs that use this API, but I do realize we don't have to limit ourselves to that. We can serve the rest of the request in a "raw" subnode of the JSON response, like this for a EC2 host:

``` json
{ 
  "power" : "on", 
  "raw" : "instancesSet" : { 
   "currentState" : { "code" : 0, "name" : "pending" },
   "previousState" : { "code" : 3, "name" : "error" },
   "instanceId" : 32329 
  }
}              
```

@mchesler  @isratrade  @ohadlevy  @domcleal  your thoughts?

**Comment by domcleal** [comment]
We don't need to worry about changing the API and getting rid of `"power": "on"` if it makes sense to improve it, since we're in the experimental APIv2 here by the way.

**Comment by dLobatog** [comment]
@domcleal right, the reason I would keep a `"power":"on"` is not actually only that, but also a way of having an easy, quick way to get this information from a host. Otherwise clients will see responses like the one above, or a mere '0' by libvirt, etc... and will have to parse and understand the response by themselves. We can provide a sort of shortcut for that to make the api more friendly.

**Comment by dLobatog** [comment]
@mchesler ping?

**Comment by mchesler** [comment]
@eLobato How much processing of the response do you want to occur here?  In theory, couldn't the response here be something like `{ 'task_state' => 'queued' }`, which would then require looking up the actual host state before responding with a specific power state?

**Comment by dLobatog** [comment]
@mchesler We want at least to respond some standard message in the `"power"` key of the JSON response, and the rest in the `"raw"` key. In the case of VMware maybe it'd require a second "status" call to be something like this?

``` json
{ 
  "power" : "on", //this is the status at the end of the call 
  "raw" : { "task_state" : "queued" }
  }
}              
```

That post-processing should ideally go under `app/services/power_manager/virt` and subclass that. Does this seem reasonable to you?

**Comment by dLobatog** [comment]
@mchesler ping?

**Comment by mchesler** [comment]
sorry, been crazy busy and haven't had time to get back to this

On Thu, Jul 17, 2014 at 11:33 AM, Daniel Lobato García <
notifications@github.com> wrote:

> @mchesler https://github.com/mchesler ping?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/theforeman/foreman/pull/1234#issuecomment-49323343.

## 

Matt Chesler
(212) 300-3971
mchesler@chesent.com

**Comment by dLobatog** [comment]
Closing this due to inactivity, and there haven't been any updates in several months. The issue is still open and I think it'd be nice if we could homogenize these responses. For the moment, @mchesler feel free to reopen if you want to keep the discussion or want to update the PR.
